include/drivers/pcie/controller.h

```diff
@@ -0,0 +1,294 @@
+ * @param reg the configuration word index (not address)
+ * @param data the value to write
+ */
+typedef void (*pcie_ctrl_conf_write_t)(const struct device *dev, pcie_bdf_t bdf,
+					  unsigned int reg, uint32_t data);
+
+/**
+ * @brief Function called to allocate a memory region subset for an endpoint Base Address Register.
+ *
+ * When enumerating PCIe Endpoints, Type0 endpoints can require up to 6 memory zones
+ * via the Base Address Registers from I/O or Memory types.
+ *
+ * This call allocates such zone in the PCI Express Controller memory regions if
+ * such region is available and space is still available.
+ *
+ * @param dev PCI Express Controller device pointer
+ * @param bdf PCI(e) endpoint
+ * @param mem True if the BAR is of memory type
+ * @param mem64 True if the BAR is of 64bit memory type
+ * @param bar_size Size in bytes of the Base Address Register as returned by HW
+ * @param bar_bus_addr bus-centric address allocated to be written in the BAR register
+ * @return True if allocation was possible, False if allocation failed
+ */
+typedef bool (*pcie_ctrl_region_allocate_t)(const struct device *dev, pcie_bdf_t bdf,
+					    bool mem, bool mem64, size_t bar_size,
+					    uintptr_t *bar_bus_addr);
+
+/**
+ * @brief Function called to translate an endpoint Base Address Register bus-centric address
+ * into Physical address.
+ *
+ * When enumerating PCIe Endpoints, Type0 endpoints can require up to 6 memory zones
+ * via the Base Address Registers from I/O or Memory types.
+ *
+ * The bus-centric address set in this BAR register is not necessarely accessible from the CPU,
+ * thus must be translated by using the PCI Express Controller memory regions translation
+ * ranges to permit mapping from the CPU.
+ *
+ * @param dev PCI Express Controller device pointer
+ * @param bdf PCI(e) endpoint
+ * @param mem True if the BAR is of memory type
+ * @param mem64 True if the BAR is of 64bit memory type
+ * @param bar_bus_addr bus-centric address written in the BAR register
+ * @param bar_addr CPU-centric address translated from the bus-centric address
+ * @return True if translation was possible, False if translation failed
+ */
+typedef bool (*pcie_ctrl_region_xlate_t)(const struct device *dev, pcie_bdf_t bdf,
+					 bool mem, bool mem64, uintptr_t bar_bus_addr,
+					 uintptr_t *bar_addr);
+
+/**
+ * @brief Read a 32-bit word from a Memory-Mapped endpoint's configuration space.
+ *
+ * Read a 32-bit word from an endpoint's configuration space from a Memory-Mapped
+ * configuration space access method, known as PCI Control Access Method (CAM) or
+ * PCIe Extended Control Access Method (ECAM).
+ *
+ * @param cfg_addr Logical address of Memory-Mapped configuration space
+ * @param bdf PCI(e) endpoint
+ * @param reg the configuration word index (not address)
+ * @return the word read (0xFFFFFFFFU if nonexistent endpoint or word)
+ */
+uint32_t generic_pcie_ctrl_conf_read(mm_reg_t cfg_addr, pcie_bdf_t bdf, unsigned int reg);
```

Comment:
It's weird to see generic_ being put before the domain's prefix. Why not pcie_generic_  ?
Unless this has been already applied like that before.

Reply:
This should be fixed on #37668 instead

Reply:
Too bad my comment was made the same day #37668 was merged. You don't seem to have made that change.

Reply:
Indeed, let me submit a PR with this name changing

Reply:
Submitted as #41948